crates/re_viewer/src/web_tools.rs

```diff
@@ -37,141 +80,278 @@ pub fn set_url_parameter_and_refresh(key: &str, value: &str) -> Result<(), wasm_
     location.assign(&url.href())
 }
 
-/// Percent-encode the given string so you can put it in a URL.
-pub fn percent_encode(s: &str) -> String {
-    format!("{}", js_sys::encode_uri_component(s))
+static STORED_HISTORY_ENTRY: OnceLock<Arc<Mutex<Option<HistoryEntry>>>> = OnceLock::new();
+
+fn stored_history_entry() -> &'static Arc<Mutex<Option<HistoryEntry>>> {
+    STORED_HISTORY_ENTRY.get_or_init(|| Arc::new(Mutex::new(None)))
+}
+
+fn get_stored_history_entry() -> Option<HistoryEntry> {
+    stored_history_entry().lock().clone()
+}
+
+fn set_stored_history_entry(entry: Option<HistoryEntry>) {
+    *stored_history_entry().lock() = entry;
+}
+
+type EventListener<Event> = dyn FnMut(Event) -> Result<(), JsValue>;
+
+/// Listen for `popstate` event, which comes when the user hits the back/forward buttons.
+///
+/// <https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event>
+pub fn install_popstate_listener(app: &mut crate::App) -> Result<(), JsValue> {
+    let egui_ctx = app.egui_ctx.clone();
+    let command_sender = app.command_sender.clone();
+
+    let closure = Closure::wrap(Box::new({
+        move |event: web_sys::PopStateEvent| {
+            let new_state = deserialize_from_state(event.state())?;
+            handle_popstate(&egui_ctx, &command_sender, new_state)?;
+            Ok(())
+        }
+    }) as Box<EventListener<_>>);
+
+    set_stored_history_entry(history()?.current_entry()?);
+
+    window()?
+        .add_event_listener_with_callback("popstate", closure.as_ref().unchecked_ref())
+        .ok_or_log_js_error();
+
+    app.popstate_listener = Some(PopstateListener(Some(closure)));
+
+    Ok(())
+}
+
+pub struct PopstateListener(Option<Closure<EventListener<web_sys::PopStateEvent>>>);
+
+impl Drop for PopstateListener {
+    fn drop(&mut self) {
+        let Some(window) = window().ok_or_log_js_error() else {
+            return;
+        };
+
+        let closure = self.0.take().unwrap();
+        window
+            .remove_event_listener_with_callback("popstate", closure.as_ref().unchecked_ref())
+            .ok_or_log_js_error();
```

Comment:
This `unwrap` can be avoided with a simple `if let Some(…)`.

We really want to avoid `unwrap` in the code since either it is a latent crash, or it requires an explanation why it is not (see https://github.com/rerun-io/rerun/issues/3408)